lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -133,55 +134,58 @@ namespace {
-                    unsigned int dataSize = primvarData.size();
-                    for (size_t v = 0; v < numVertices; v++) {
+            if (numVertices <= renderingToSceneFaceVtxIds.size()) {
+                const unsigned int dataSize = primvarData.size();
+                for (size_t v = 0; v < numVertices; v++) {
+                    unsigned int index = renderingToSceneFaceVtxIds[v];
+                    if (index < dataSize) {
                         SRC_TYPE* pointer = reinterpret_cast<SRC_TYPE*>(
                             reinterpret_cast<float*>(&vertexBuffer[v]) + channelOffset);
-                        unsigned int index = faceVertexIndices[v];
-                        if (index < dataSize) {
-                            *pointer = primvarData[index];
-                        } else {
-                            TF_DEBUG(HDVP2_DEBUG_MESH).Msg("Invalid Hydra prim '%s': "
-                                                    "primvar %s has %u elements, while its topology "
-                                                    "references face vertex index %u.\n",
-                                                    rprimId.asChar(), primvarName.GetText(),
-                                                    dataSize, index);
-                        }
+                        *pointer = primvarData[index];
+                    }
+                    else {
+                        TF_DEBUG(HDVP2_DEBUG_MESH).Msg("Invalid Hydra prim '%s': "
+                            "primvar %s has %u elements, while its topology "
+                            "references face vertex index %u.\n",
+                            rprimId.asChar(), primvarName.GetText(),
+                            dataSize, index);
                     }
                 }
-                else {
-                    // numVertices must have been assigned with the number of
-                    // face vertices as required by vertex unsharing.
-                    TF_CODING_ERROR("Invalid Hydra prim '%s': primvar %s "
-                        "requires %zu unshared elements, while the number of "
-                        "face vertices is %zu. Skipping primvar update.",
-                        rprimId.asChar(), primvarName.GetText(),
-                        numVertices, faceVertexIndices.size());
-                }
             }
-            else if (numVertices <= primvarData.size()) {
+            else {
+                TF_CODING_ERROR("Invalid Hydra prim '%s': "
+                    "requires %zu vertices, while the number of elements in "
+                    "renderingToSceneFaceVtxIds is %zu. Skipping primvar update.",
+                    rprimId.asChar(), numVertices, renderingToSceneFaceVtxIds.size());
+
+                memset(vertexBuffer, 0, sizeof(DEST_TYPE) * numVertices);
+            }
+            break;
+        case HdInterpolationUniform:
+        {
+            // Unshared vertex layout is required for uniform primvars, in this case
```

Comment:
Can we assert here that renderingToSceneVertexIds is holding what we think it is?

Reply:
Hmm... That would be a linear loop to assert its element values? I don't see big values of this assertion, maybe if we hit issue we could add back the _requiresUnsharedVertices for the verification purpose.

Reply:
I want something to catch the case where someone modifies how the buffer is generated for the unshared case, but then don't realize they need to update this algorithm because the data isn't used. I don't think we're going to hit bad data and that will cause a problem, I think we'll potentially modify the code and then not understand why things don't work.

It doesn't have to be a 100% verification, we could do something like in debug builds test every 1000th value and make sure it is what we expect. Probably that'll catch the case I'm worried about.

Reply:
I think this makes sense. Maybe for the moment I could add a comment to point to the other code area when changing the array data.

Reply:
But on the other hand, I don't see how we would change the new face vertex ids for unsharing case...